src/coq_elpi_builtins.ml

```diff
@@ -158,14 +160,23 @@ let add_universe_constraint state c =
 let mk_fresh_univ state = new_univ state
 
 let mk_algebraic_super x = Sorts.super x
-let mk_algebraic_max x y = Sorts.sort_of_univ (Univ.Universe.sup (Sorts.univ_of_sort x) (Sorts.univ_of_sort y))
+let mk_algebraic_max x y =
+  let open Sorts in
+  match x, y with
+  | (SProp, SProp) | (Prop, Prop) | (Set, Set) -> x
+  | (SProp, (Prop | Set | Type _)) | ((Prop | Set | Type _), SProp) -> raise Pred.No_clause
+  | (Prop, (Set | Type _ as u)) | ((Set | Type _ as u), Prop) -> u
+  | (Set, Type u) | (Type u, Set) -> Sorts.sort_of_univ (Univ.Universe.sup u Univ.Universe.type0)
+  | (Type u, Type v) -> Sorts.sort_of_univ (Univ.Universe.sup u v)
```

Comment:
@ppedrot I'm not so fond of having CIC typing rules in Elpi. These should not be here.

Reply:
The max of two levels doesn't exist either. It would be a `Universe.t`, i.e. some free algebraic expression. Elpi is confusing the three notions of level (a unique name), algebraic (list of levels + integer increments) and sorts (Prop, SProp, and Type@@{alg}). Either you expose the true complexity of the various objects to the user, or you hide it under a palatable interface based on sorts (the only real thing the user has access to in Coq) and you don't have max anymore but more refined notions. (By the way, sup is a lie as well, e.g. Prop + 1 is not Prop + 1 but Set + 1. This function computes the type of a sort.)

Reply:
I think I need to chat with you next week. I agree with what you say, but I wanted to give anothe illusion...

Reply:
No problem. For now I don't have many constraints except Tuesday afternoon and the weekly call.